File: leetcode/contest-254/1970. Last Day Where You Can Still Cross.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
   public:
    unordered_map<int, unordered_map<int, int>> mp;
    int row, col;
    unordered_map<int, unordered_map<int, int>> f;
    unordered_map<int, unordered_map<int, int>> v;
    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};
    bool dfs(int i, int j, int m) {
        if (f.count(i) && f[i].count(j)) {
            return f[i][j];
        }
        // 如果这个点变成蓝色了
        if (mp.count(i) && mp[i].count(j) && mp[i][j] < m) {
            return f[i][j] = false;
        }
        if (i == row - 1) {
            return f[i][j] = true;
        }
        v[i][j] = true;
        for (int k = 0; k < 4; k++) {
            int x = dx[k] + i, y = dy[k] + j;
            if (x >= 0 && x < row && y >= 0 && y < col) {
                // 不是直接用 v[x][y] 可以加快速度，否则 TLE
                if (!(v.count(x) && v[x].count(y) && v[x][y]) && dfs(x, y, m)) {
                    v[i][j] = false;
                    return f[i][j] = true;
                }
            }
        }
        v[i][j] = false;
        return f[i][j] = false;
    }
    bool valid(int m) {
        f.clear();
        v.clear();

        for (int j = 0; j < col; j++) {
            // cout << j << " " << m << endl;
            if (dfs(0, j, m)) return true;
        }
        return false;
    }
    int latestDayToCross(int _row, int _col, vector<vector<int>>& cells) {
        for (int i = 0; i < cells.size(); i++) {
            auto& c = cells[i];
            mp[c[0] - 1][c[1] - 1] = i;
        }
        row = _row, col = _col;
        int l = 0, r = row * col;
        // 找最大的 valid
        while (l < r) {
            int mid = (l + r + 1) >> 1;
            if (valid(mid))
                l = mid;
            else
                r = mid - 1;
        }
        return l;
    }
};

File: leetcode/残酷刷题/1970. Last Day Where You Can Still Cross.ts
function latestDayToCross (row: number, col: number, cells: number[][]): number {
  const [s, t] = [0, row * col + 1]
  const fa = [...Array<number>(row * col + 2)].map((_, idx) => idx)
  function get (x: number): number {
    if (x === fa[x]) return x
    return fa[x] = get(fa[x])
  }
  function merge (x: number, y: number) {
  // console.log(x, y)
    const fx = get(x); const fy = get(y)
    fa[fx] = fy
  }
  const idx = (r: number, c: number) => (r - 1) * col + c
  for (let j = 1; j <= col; j++) {
    merge(s, idx(1, j))
    merge(t, idx(row, j))
  }
  const f = Array<boolean>(row * col + 2).fill(false)
  const ok = (i: number, j: number) => i >= 1 && i <= row && j >= 1 && j <= col && f[idx(i, j)]
  for (let k = cells.length - 1; k >= 0; k--) {
    // console.log('get', get(s), get(t))
    if (get(s) === get(t)) return k + 1

    const [r, c] = cells[k]
    // console.log('cell', r, c)
    f[idx(r, c)] = true
    if (ok(r - 1, c))
    { merge(idx(r - 1, c), idx(r, c)) }
    if (ok(r + 1, c))
    { merge(idx(r + 1, c), idx(r, c)) }
    if (ok(r, c + 1))
    { merge(idx(r, c + 1), idx(r, c)) }
    if (ok(r, c - 1))
    { merge(idx(r, c - 1), idx(r, c)) }
  }
  return 0
};
